File: src/main/java/wraith/fabricaeexnihilo/json/recipe/FluidOnTopRecipeJson.java
package wraith.fabricaeexnihilo.json.recipe;

import com.google.gson.*;
import wraith.fabricaeexnihilo.api.crafting.FluidIngredient;
import wraith.fabricaeexnihilo.api.recipes.barrel.FluidOnTopRecipe;
import wraith.fabricaeexnihilo.json.BaseJson;
import wraith.fabricaeexnihilo.modules.barrels.modes.BarrelMode;
import wraith.fabricaeexnihilo.modules.barrels.modes.EmptyMode;

import java.lang.reflect.Type;

public final class FluidOnTopRecipeJson extends BaseJson<FluidOnTopRecipe> {

    private FluidOnTopRecipeJson() {}

    public static final FluidOnTopRecipeJson INSTANCE = new FluidOnTopRecipeJson();

    @Override
    public FluidOnTopRecipe deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        var obj = json.getAsJsonObject();
        return new FluidOnTopRecipe(
                FluidIngredient.fromJson(obj.get("inBarrel"), context),
                FluidIngredient.fromJson(obj.get("onTop"), context),
                BarrelMode.of(obj.get("result"), context)
        );
    }

    @Override
    public JsonElement serialize(FluidOnTopRecipe src, Type typeOfSrc, JsonSerializationContext context) {
        var obj = new JsonObject();
        obj.add("inBarrel", src.inBarrel().toJson(context));
        obj.add("onTop", src.onTop().toJson(context));
        if(!(src.result() instanceof EmptyMode)) {
            var result = new JsonObject();
            result.add(src.result().nbtKey(), context.serialize(src.result()));
            obj.add("result", result);
        }
        return obj;
    }

}

File: src/main/kotlin/exnihilofabrico/json/FluidOnTopRecipeJson.kt
package exnihilofabrico.json

import com.google.gson.*
import exnihilofabrico.api.crafting.FluidIngredient
import exnihilofabrico.api.recipes.barrel.FluidOnTopRecipe
import exnihilofabrico.modules.barrels.modes.EmptyMode
import exnihilofabrico.modules.barrels.modes.barrelModeFactory
import java.lang.reflect.Type

object FluidOnTopRecipeJson: JsonDeserializer<FluidOnTopRecipe>, JsonSerializer<FluidOnTopRecipe> {
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): FluidOnTopRecipe {
        val obj = json.asJsonObject
        return FluidOnTopRecipe(
            FluidIngredient.fromJson(obj["inBarrel"], context),
            FluidIngredient.fromJson(obj["onTop"], context),
            barrelModeFactory(obj["result"], context)
        )
    }

    override fun serialize(src: FluidOnTopRecipe, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val obj = JsonObject()
        obj.add("inBarrel", src.inBarrel.toJson(context))
        obj.add("onTop", src.onTop.toJson(context))
        if(src.result !is EmptyMode) {
            val result = JsonObject()
            result.add(src.result.tagKey(), context.serialize(src.result))
            obj.add("result", result)
        }
        return obj
    }

}